scipy/optimize/_canonical_constraint.py

```diff
@@ -0,0 +1,480 @@
+from __future__ import division, print_function, absolute_import
+import numpy as np
+import scipy.sparse as spc
+from ._constraints import (NonlinearConstraint,
+                           LinearConstraint,
+                           BoxConstraint)
+
+__all__ = ['CanonicalConstraint',
+           'to_canonical',
+           'lagrangian_hessian',
+           'empty_canonical_constraint']
+
+
+class CanonicalConstraint:
```

Comment:
A general thought: could it be better to split this constraint into equality and inequality parts? Are there any benefits to keep them together (seems sort of unnatural)?

Reply:
Actually, I choose to implement it like this  because of constraints of the type 
``lb <= fun(x) <= ub``, which can specify, simultaneously, equality (by setting ``lb[i] == ub[i]``) and inequality constraints. 

So specify the canonical constraint like this actually save us from a lot of trouble. I tried to implement them separately at first and everything gets much more complicated.